### ISSUE TITLE
Incorporating pooled spam & ham with model downloads

### ISSUE THREAD

**Issue by Ferrariic** [jackriccomini/spamfilter-plugin-runelite#1]
Excited to see this project. 

My concern is that it may take a user a very long time to accrue the necessary training data, spam & ham in this case - and models may be wildly different in efficacy.

An alternative option would be to send 'Spam & Ham' text to a mysql database, and then train a model which is then loaded onto the plugin via a 'Sync' command. You could run the plugin's sync feature once a day, or allow configuration in the user settings. 

Effectively, the most reported phrases & most normal flagged phrases would be used in training, perhaps the upper 50%?

Best of luck with this project, it's very exciting.

**Comment by jackriccomini** [comment]
Thanks for the suggestion! This is something I've thought about before and I agree that it'd be useful. Here's my thinking:

> My concern is that it may take a user a very long time to accrue the necessary training data

It's very true that more data is better.

My intuition would be that in the cat-and-mouse game of spam detection, your options are to either build filters that are effective against new unseen spam or to build filters which can quickly adapt to new unseen spam. The former would be helped by more data. The latter is thankfully already pretty easy with a Bayesian filter; experimentally it seems like a message that gets through the filter can often be blocked with one or two "mark as spam"s. This means that, while I completely agree that more data would be useful, it's also something that can be put off for a while.

If enough people end up playing with the plugin then this is definitely on the cards! For now though, there are a number of simpler tweaks to be made to the local filtering which would probably be a higher priority.

> models may be wildly different in efficacy.

It's not a bug, it's a feature™.
One of the nice things about training a local model is that you can decide for yourself what constitutes spam.
For example, I included the message "Selling bulk runes and dart tips" in the built-in spam corpus.
I had no reason to believe that the message came from a bot, I just wasn't interested in seeing it.

The most correct way to handle this with user-submitted spam data would be to have a number of different corpuses which can be reported to and then selectively enabled in the settings, e.g. "bot", "trading", "clan recruiting", "scammer".
The simplest way would be to lump everything in together but to have a high criteria for acceptance so that only the most agreed upon spam messages are counted as spam.

> Best of luck with this project, it's very exciting.

Thanks!

You should know that right now this is all hypothetical future work. I hacked this plugin together for my own use a few weeks ago before I let my membership lapse. I cleaned it up the other day and submitted it to the plugin hub on the off chance that someone would find it useful but I didn't anticipate that it would end up getting any great amount of interest. Right now I'm mostly able to commit to bug fixes and the occasional simple feature if requested but anything bigger will have to wait until I play more regularly.

**Comment by jackriccomini** [comment]
To clarify a point above, there's nothing inherent about user-submitted reports which would reduce the effectiveness of local reporting. Right now if someone disagrees with my built-in assessments of spam and ham then they meet the same issue. The difference is that with the current small amount of data out of the box, the built-in model is pretty easy to override with only a small number of user reports. Fetching a large data set OTOH would naturally mean that the effect of user reports would be diluted and so the filter would need to have a very large bias in favour of user reports if the user wants to override the built-in data

**Comment by extreme4all** [comment]
just wondering, is this just counting the number of good words and bad words in the corpus? as a prediction, or am i reading the code wrong?

**Comment by jackriccomini** [comment]
> just wondering, is this just counting the number of good words and bad words in the corpus? as a prediction, or am i reading the code wrong?

No, it's not just counting. It is unusually simple though for what technically qualifies as "machine learning". I based my implementation on [this](http://www.paulgraham.com/spam.html) article. A more thorough explanation is [here](http://cs.wellesley.edu/~anderson/writing/naive-bayes.pdf). The quickest explanation is:

1. Calculate the probability that each token (word) is spam by counting how many times it appears in each corpus. Unknown words are assumed to have 40% probability of being spam
2. Use Bayes' rule to combine those probabilities into a probability that the entire message is spam

**Comment by extreme4all** [comment]
@jackriccomini are you still actively maintaining this project?

if you are i'd like to bounce some ideas for improvements. You can contact me on discord via: extreme4all#6456

**Comment by jackriccomini** [comment]
> @jackriccomini are you still actively maintaining this project?
> 
> if you are i'd like to bounce some ideas for improvements. You can contact me on discord via: extreme4all#6456

Yes. I don't play much recently but when I do play I use the plugin so as long as it doesn't break too often I intend to keep it maintained for my own use (and if it does break too often during periods that I'm unwilling to commit time to maintenance, I'll remove the plugin from RL's plugin hub). I don't need much in the way of features - as long as it's working, it's good enough for me with the current small feature set. If I play more regularly in the future then my needs might change but at least for now new improvements aren't much of a priority for me.

While I do appreciate the interest in wanting to improve the plugin, I don't really use Discord. I'm always happy to entertain suggestions and issues filed here on GitHub. Even if what somebody suggests isn't relevant to what I need out of a spam filter, anybody would be welcome to implement it themself and either fork the plugin or send a pull request to me.

During weeks or months that I don't actively play I will try to resolve crashes or major bugs reported to me.
